File: source/math/ortho_normal_basis.cpp
#include "math/ortho_nomal_basis.hpp"
#include "math/vec.hpp"

#include <cmath>

ortho_normal_basis::ortho_normal_basis(const norm &x_in, const norm &y_in, const norm &z_in)
{
   data[0] = x_in;
   data[1] = y_in;
   data[2] = z_in;
}

ortho_normal_basis ortho_normal_basis::from_xy(const norm &x_in, const norm &y_in)
{
   const norm z = normalise(cross(x_in, y_in));
   const norm y = cross(z, x_in);

   return ortho_normal_basis{x_in, y, z};
}

ortho_normal_basis ortho_normal_basis::from_yx(const norm &y_in, const norm &x_in)
{
   const norm z = normalise(cross(x_in, y_in));
   const norm x = cross(y_in, z);

   return ortho_normal_basis{x, y_in, z};
}

ortho_normal_basis ortho_normal_basis::from_xz(const norm &x_in, const norm &z_in)
{
   const norm y = normalise(cross(z_in, x_in));
   const norm z = cross(x_in, y);

   return ortho_normal_basis{x_in, y, z};
}

ortho_normal_basis ortho_normal_basis::from_zx(const norm &z_in, const norm &x_in)
{
   const norm y = normalise(cross(z_in, x_in));
   const norm x = cross(y, z_in);

   return ortho_normal_basis{x, y, z_in};
}

ortho_normal_basis ortho_normal_basis::from_yz(const norm &y_in, const norm &z_in)
{
   const norm x = normalise(cross(y_in, z_in));
   const norm z = cross(x, y_in);

   return ortho_normal_basis{x, y_in, z};
}

ortho_normal_basis ortho_normal_basis::from_zy(const norm &z_in, const norm &y_in)
{
   const norm x = normalise(cross(y_in, z_in));
   const norm y = cross(x, z_in);

   return ortho_normal_basis{x, y, z_in};
}

ortho_normal_basis ortho_normal_basis::from_z(const norm &z_in)
{
   const norm x = normalise(cross(
      std::fabs(dot(z_in, norm(1.0, 0.0, 0.0))) > 0.999 ? norm(0.0, 1.0, 0.0) : norm(1.0, 0.0, 0.0),
      z_in));
   const norm y = normalise(cross(z_in, x));

   return ortho_normal_basis{x, y, z_in};
}

File: go/maths/ortho_normal_basis.go
package maths

import "math"

type OrthoNormalBasis struct {
   X Vec3
   Y Vec3
   Z Vec3
}

func (lhs *OrthoNormalBasis) Transform(rhs *Vec3) Vec3 {
   x := MultScalar(&lhs.X, rhs.X)
   y := MultScalar(&lhs.Y, rhs.Y)
   z := MultScalar(&lhs.Z, rhs.Z)

   xy := Add(&x, &y);

   return Add(&xy, &z)
}

func FromXY(x *Vec3, y *Vec3) OrthoNormalBasis {
   crossXY := Cross(x, y)
   normZ := Normalise(&crossXY)
   normY := Cross(&normZ, x)
   
   return OrthoNormalBasis{X: *x, Y: normY, Z: normZ}
}

func FromYX(y *Vec3, x *Vec3) OrthoNormalBasis {
    crossXY := Cross(x, y)
    normZ := Normalise(&crossXY)
    normX := Cross(y, &normZ)

    return OrthoNormalBasis{X: normX, Y: *y, Z: normZ}
}

func FromXZ(x *Vec3, z *Vec3) OrthoNormalBasis {
    crossZX := Cross(z, x)
    normY := Normalise(&crossZX)
    normZ := Cross(x, &normY)

    return OrthoNormalBasis{X: *x, Y: normY, Z: normZ}
}

func FromZX(z *Vec3, x *Vec3) OrthoNormalBasis {
    crossZX := Cross(z, x)
    normY := Normalise(&crossZX)
    normX := Cross(&normY, z)

    return OrthoNormalBasis{X: normX, Y: normY, Z: *z}
}

func FromYZ(y *Vec3, z *Vec3) OrthoNormalBasis {
    crossYZ := Cross(y, z) 
    normX := Normalise(&crossYZ)
    normZ := Cross(&normX, y)

    return OrthoNormalBasis{X: normX, Y: *y, Z: normZ}
}

func FromZY(z *Vec3, y *Vec3) OrthoNormalBasis {
    crossYZ := Cross(y, z)
    normX := Normalise(&crossYZ)
    normY := Cross(&normX, z)

    return OrthoNormalBasis{X: normX, Y: normY, Z: *z}
}

func FromZ(z* Vec3) OrthoNormalBasis {
    xAxis := Vec3{X: 1.0}

    if math.Abs(Dot(z, &xAxis)) >= 0.999 {
        basis := Vec3{Y: 1.0}
        crossBZ := Cross(&basis, z)
        normX := Normalise(&crossBZ)
        
        crossZX := Cross(z, &normX)
        normY := Normalise(&crossZX)

        return OrthoNormalBasis{X: normX, Y: normY, Z: *z}

    } else {
        basis := xAxis
        crossBZ := Cross(&basis, z)
        normX := Normalise(&crossBZ)
        
        crossZX := Cross(z, &normX)
        normY := Normalise(&crossZX) 

        return OrthoNormalBasis{X: normX, Y: normY, Z: *z}
    }
}
